### PR TITLE
Fixed nested italics (#4)

### PR DESCRIPTION
**Pull request by tillig**
Fix https://github.com/gitterHQ/marked/issues/4

This at least partially fixes the issues reported in (https://github.com/gitterHQ/marked/issues/4). It doesn't fix the use case

:chart_with_upwards_trend: followed by _emphasized text_

But it does fix the other nested italics/emphasis issues. Includes additional tests to verify functionality.

### CODE REVIEW

File: test/tests/nested_em.text [hunk 1]
```diff
@@ -1,3 +1,15 @@
 *test **test** test*
 
 _test __test__ test_
+
+_test_ test _test_
```
Comment (MadLittleMods) [inline test/tests/nested_em.text]: Put these in their own test `multiple_em`

File: test/tests/nested_em.text [hunk 2]
```diff
@@ -1,3 +1,15 @@
...
+
+*test* test *test*
+
+_test_ __test__ _test_
+
+*test* **test** *test*
+
+_a_ _b_
+
+*a* *b*
```
Comment (MadLittleMods) [inline test/tests/nested_em.text]: Let's add some more tests

 - `\*foo *bar*` -> `<p>*foo <em>bar</em></p>` -> \*foo *bar*
 - `\_foo _bar_` -> `<p>*foo <em>bar</em></p>` -> \_foo _bar_
 - `\*foo *bar* and *baz*` -> `<p>*foo <em>bar</em> and <em>baz</em></p>` -> \*foo *bar* and *baz*
 - `\_foo _bar_ and _baz_` -> `<p>*foo <em>bar</em> and <em>baz</em></p>` -> \_foo _bar_ and _baz_

**Comment by tillig** [comment]
With respect to the new tests desired...

`\*foo *bar*` -> `<p>*foo <em>bar</em></p>` -> \*foo *bar*
`\_foo _bar_` -> `<p>*foo <em>bar</em></p>` -> \_foo _bar_
`\*foo *bar* and *baz*` -> `<p>*foo <em>bar</em> and <em>baz</em></p>` -> \*foo *bar* and *baz*
`\_foo _bar_ and _baz_` -> `<p>*foo <em>bar</em> and <em>baz</em></p>` -> \_foo _bar_ and _baz_

Do you want those in `nested_em.text` or `multiple_em.text`? Or some other new fixture?

**Comment by MadLittleMods** [comment]
Let's put the first two in a new `no_trample_em.text` test file. The last two can go in `multiple_em.text`

**Comment by tillig** [comment]
New changes are in. Hopefully I put the new tests (which do pass) into the correct fixtures.

**Comment by MadLittleMods** [comment]
The latest from `marked` seems to solve all the issues in #4 

https://github.com/chjj/marked/blob/88ce4df47c4d994dc1b1df1477a21fb893e11ddc/lib/marked.js#L459
```
^\b_((?:[^_]|__)+?)_\b|^\*((?:\*\*|[\s\S])+?)\*(?!\*)
```

---

We can then also add this test to `no_trample_em`

 - `:chart_with_upwards_trend: followed by _emphasized text_` -> `<p>:chart_with_upwards_trend: followed by <em>emphasized text</em></p>`

**Comment by tillig** [comment]
How exactly would you like me to proceed here? I want to get you the right fix for an issue that seems to have been open for quite some time with no traction, but it also is becoming potentially a little overly iterative. For example, the extra tests could have been added post-PR. I'm happy to put in the new regex as outlined above, but I'm also afraid of this turning into scope creep.

If I get the new regex in and the additional test can we call it good?

**Comment by MadLittleMods** [comment]
@tillig Yep, I was just re-reading the PR and reviewing fully and noticed the latest regex upstream solved all issues. Sorry for the iteration. I wanted the extra tests to ensure we aren't breaking things as we move along. This code can have a widespread effects for future messages.

**Comment by tillig** [comment]
Updated with the new regex and test as specified.

**Comment by MadLittleMods** [comment]
@tillig Thanks for the contribution :grinning: 

I'll make a release for this package, bump the dep in the Gitter app, and it will come out in upcoming release.

**Comment by suprememoocow** [comment]
@tillig Thank you Travis!

**Pull request merged**